RearrangeCarGroup8Movements.py

```diff
@@ -0,0 +1,195 @@
+        self.car = car
+        self.from_index = from_index
+        self.to_index = to_index
+
+    def __str__(self):
+        move_str = "Move car #%s from %s to %s" %(self.car, self.from_index, self.to_index)
+        return move_str
+
+    def print_log_file(self):
+        move_str = "Move car #%s from %s to %s" %(self.car, self.from_index, self.to_index)
+        logging.info(move_str)
+
+class RearrangeCars(object):
+
+    '''Class to rearrange the cars and print the moves
+       - Uses car_position_dict to optimize searching the position of a given car
+       - Stores additionally the position of empty spot in current car arrangement
+       self.current_car_order.
+       - Moves are either printed to the console or logged
+       - Stores and updates min_num_steps_left that shows min number of moves left
+        in order to avoid comparison current and final order lists that takes O(n)
+    '''
+
+    def __init__(self, init_car_order, final_car_order):
+
+        '''Initializes class instance with initial, final order and
+        stores the position of empty parking space in the current car order
+        Memory is O(n) to store 3 order lists, 1 car_position_dict and a set of
+        not arranged cars
+
+        Storing car_position_dict helps to search indices of car in O(1) instead of O(n)
+        A set of not_arranged_cars helps to now when to stop in O(1) as well as to draw
+        in O(1) unarranged car
+
+        Arguments:
+           init_car_order: a valid non-empty list of cars + 1 \
+           empty_place given in their parking place order
+
+           final_car_order: a valid non-empty list of cars + 1 \
+           empty_place according to the desired order
+
+        Returns:
+            None
+        '''
+
+        self.initial_car_order = init_car_order
+        self.current_car_order = init_car_order
+        self.final_car_order = final_car_order
+        self.empty_position = self.current_car_order.index(0)
+        self.print_console = True
+        self.car_position_dict = {car:pos for pos, car in enumerate(init_car_order) \
+                                  if car} #search the index of car in O(1) via dict
+        self.not_arranged_cars = {car for i, car in enumerate(init_car_order)\
+                                  if init_car_order[i] != final_car_order[i] and car}
+
+    def _create_log_file(self):
+
+        '''In case the number of moves is more than 1000
+        writes the output to the logfile instead of the console
+        '''
+
+        num_cars = len(self.not_arranged_cars)
+        info_message = 'Number of cars to arrange is {0} > 1000, movements are logged'.format(num_cars)
+        logging_message = info_message
+        logging.warning(logging_message)
+        logging.info('Initial car order:')
+        logging.info(self.initial_car_order)
+        logging.info('Final car order:')
+        logging.info(self.final_car_order)
+        logging.info('Moves:')

```

Comment:
I have never used `logging` before, so I am a bit confused... When I was testing from my computer, I couldn't really find the log file, but I saw all the logging info being printed out to the console...